### PR TITLE
[sat] Add option to disable Minisat simplifications

### PR DESCRIPTION
**Pull request by HanielB**
Useful to better control behavior, specially in terms of how the solver performs
when these simplifications (mainly variable and clause elimination) are
disabled, e.g. with incremental and/or proofs.

Also renames the previously confusing "--minisat--elimination" option, which
only refers to variable elimination, a subset of the Minisat simplifications.

Moreover deletes an old comment which no longer applies.

### CODE REVIEW

File: src/smt/set_defaults.cpp [hunk 1]
```diff
@@ -684,10 +684,9 @@ void SetDefaults::setDefaultsPost(const LogicInfo& logic, Options& opts) const
     }
   }
 
-  // until bugs 371,431 are fixed
-  if (!opts.prop.minisatUseElimWasSetByUser)
+  if (!opts.prop.minisatVarElimWasSetByUser)
```
Comment (aniemetz) [inline src/smt/set_defaults.cpp:687]: Don't we also have to guard for `minisatSimp` here?

Comment (HanielB) [reply src/smt/set_defaults.cpp:687]: No. The "varElim" option was added (it does not exist on original Minisat) specifically to allow all other simplifications except variable elimination, which is problematic for models and the other stuff in this block.

Comment (aniemetz) [reply src/smt/set_defaults.cpp:687]: The description of `minisatSimp` says `use Minisat variable and clause elimination, plus other simplifications` though?

Comment (HanielB) [reply src/smt/set_defaults.cpp:687]: Yes. So the behavior is that if `minisatSimp` *and* `minisatVarElim` are on, the solver will do variable and clause elimination and some other stuff. If `minisatSimp` is off, does nothing. If `minisatSimp` is on but `minisatVarElim` is off, does all the simplifications except for variable elimination.

Changing the helper of `minisatVarElim` to `"use Minisat variable elimination (requires --minisat-simplification)" would help? It'd not account for the disabling of var elim even when `minisatSimp` is on, though, but I think it'd get convoluted to explain this is the helper.

Comment (aniemetz) [reply src/smt/set_defaults.cpp:687]: So I think this should be made more clear in both help messages. Right now it is implying a very different behavior (them being independent, and both enabling variable elimination).

Comment (HanielB) [reply src/smt/set_defaults.cpp:687]: I gave it a shot, PTAL

File: src/options/prop_options.toml [hunk 2]
```diff
@@ -64,12 +64,20 @@ name   = "SAT Layer"
   help       = "refine theory conflict clauses (default false)"
 
 [[option]]
-  name       = "minisatUseElim"
+  name       = "minisatVarElim"
```
Comment (ajreynol) [inline src/options/prop_options.toml:67]: Can we just make all of this a single, 3 mode option?  Having 2 options here seems more confusing.

**Review by ajreynol — CHANGES_REQUESTED** [review]

File: src/options/prop_options.toml (hunk 2 above)
Comment (HanielB) [reply src/options/prop_options.toml:67]: Good point. I'll do that.

File: src/prop/minisat/simp/SimpSolver.cc [hunk 3]
```diff
@@ -62,27 +61,18 @@ SimpSolver::SimpSolver(Env& env,
...
-      use_elim(options().prop.minisatUseElim && !enableIncremental),
       merges(0),
       asymm_lits(0),
       eliminated_vars(0),
       elimorder(1),
-      use_simplification(!enableIncremental && !options().smt.unsatCores
-                         && !pnm)  // TODO: turn off simplifications if
-                                   // proofs are on initially
-      ,
+      use_simplification(
```
Comment (ajreynol) [inline src/prop/minisat/simp/SimpSolver.cc:68]: We should not be passing `pnm` to SimpSolver, nor checking whether it is non-null here. I think we need to change `!pnm` in this check to `!env.isSatProofProducing()`.

**Review by ajreynol — CHANGES_REQUESTED** [review]

File: src/options/prop_options.toml [hunk 4]
```diff
@@ -64,12 +64,22 @@ name   = "SAT Layer"
...
-  type       = "bool"
-  default    = "true"
-  help       = "use Minisat elimination"
+  long       = "minisat-simplification=MODE"
+  type       = "MinisatSimpMode"
+  default    = "ALL"
+  help       = "Simplifications to be performed by Minisat."
+  help_mode  = "Modes for Minisat simplifications."
+[[option.mode.ALL]]
+  name = "status"
```
Comment (ajreynol) [inline src/options/prop_options.toml:75]: Fix name.

File: src/options/prop_options.toml [hunk 5]
```diff
@@ -64,12 +64,22 @@ name   = "SAT Layer"
...
+  help = "Variable and clause elimination, plus other simplifications."
+[[option.mode.NO_VAR_ELIM]]
+  name = "no-var-elim"
```
Comment (ajreynol) [inline src/options/prop_options.toml:78]: Maybe call this `clause-elim` and the mode `CLAUSE_ELIM`.

File: src/prop/minisat/simp/SimpSolver.cc (hunk 3 above)
Comment (ajreynol) [reply src/prop/minisat/simp/SimpSolver.cc:68]: Ping.

Would also help to make this not check for incremental here.

Comment (HanielB) [reply src/prop/minisat/simp/SimpSolver.cc:68]: This requires changing several interfaces, I think it's better to be done in anather PR.

Comment (ajreynol) [reply src/prop/minisat/simp/SimpSolver.cc:68]: Ok

**Review by ajreynol — APPROVED** [review]

**Pull request merged**